crates/api/src/agent.rs

```diff
@@ -0,0 +1,362 @@
+//! #### Json Schemas
+//!
+//! ```json
+//! {
+//!   "title": "AgentInfoSigned",
+//!   "type": "object",
+//!   "properties": {
+//!     "agentInfo": { "type": "string", "required": true, "description": "json AgentInfo" },
+//!     "signature": { "type": "string", "required": true, "description": "base64" }
+//!   }
+//! }
+//! ```
+//!
+//! ```json
+//! {
+//!   "title": "AgentInfo",
+//!   "type": "object",
+//!   "properties": {
+//!     "agent": { "type": "string", "required": true, "description": "base64" },
+//!     "space": { "type": "string", "required": true, "description": "base64" },
+//!     "createdAt": {
+//!         "type": "string",
+//!         "required": true,
+//!         "description": "i64 micros since unix epoch"
+//!     },
+//!     "expiresAt": {
+//!         "type": "string",
+//!         "required": true,
+//!         "description": "i64 micros since unix epoch",
+//!     },
+//!     "isTombstone": { "type": "boolean", "required": true },
+//!     "url": { "type": "string", "description": "optional" },
+//!     "storageArc": {
+//!       "type": "array",
+//!       "description": "optional",
+//!       "items": [
+//!         {
+//!             "type": "number",
+//!             "required": true,
+//!             "description": "u32 arc start loc"
+//!         },
+//!         {
+//!             "type": "number",
+//!             "required": true,
+//!             "description": "u32 arc end loc"
+//!         }
+//!       ]
+//!     }
+//!   }
+//! }
+//! ```
+//!
+//! #### Cryptography
+//!
+//! This module and its data structures are designed to be agnostic to
+//! cryptography. It exposes the [Signer] and [Verifier] traits to allow
+//! implementors to choose the algorithm to be used.
+//!
+//! The underlying data structures, however, cannot be quite so agnostic.
+//!
+//! By convention, absent other indications, the [AgentInfo::agent] property
+//! will be an ed25519 public key, and the [AgentInfoSigned::get_signature] will
+//! be an ed25519 signature.
+//!
+//! Future versions of this library may look for an optional "alg" property
+//! on the [AgentInfo] type before falling back to this usage of ed25519.
+//! These other algorithms may treat the [AgentInfo::agent] property as a
+//! hash of the public key instead of the public key itself, and find the
+//! public key instead on an "algPubKey" property. (Some post-quantum
+//! algorithms have ridiculously long key material.)
+
+use crate::*;
+
+/// Defines a type capable of cryptographic signatures.
+pub trait Signer {
+    /// Sign the encoded data, returning the resulting detached signature bytes.
+    fn sign(
+        &self,
+        agent_info: &AgentInfo,
+        message: &[u8],
+    ) -> BoxFut<'_, std::io::Result<bytes::Bytes>>;
+}
+
+/// Defines a type capable of cryptographic verification.
+pub trait Verifier {
+    /// Verify the provided detached signature over the provided message.
+    /// Returns `true` if the signature is valid.
+    fn verify(
```

Comment:
Should verify return a `Result<bool, std::io::Error>`? Do we want to bury errors by treating them as equivalent to failed verification?

Reply:
As far as I understand the algorithm, all the errors that *can* happen *are* equivalent to the signature not being valid... Namely bad-sig-size and bad-agent-pub-key.